lib/experimental/Forms/AvatarNameSelector/AvatarNameSelectorContent/AvatarNameSelectorMainContent/index.tsx

```diff
@@ -319,6 +319,48 @@ export const AvatarNameSelectorMainContent = ({
     ]
   )
 
+  const [allVisibleSelected, anyVisibleSelected] = useMemo(() => {
+    if (!entities.length) {
+      return [false, false]
+    }
+    const selectedMap = new Map<number, AvatarNamedEntity>(
+      (selectedEntities ?? []).map((se) => [se.id, se])
+    )
+
+    let visibleCount = 0
+    let selectedVisibleCount = 0
+
+    if (!groupView) {
+      visibleCount = entities.length
+      entities.forEach((fe) => {
+        if (selectedMap.has(fe.id)) {
+          selectedVisibleCount += 1
+        }
+      })
+    } else {
+      entities.forEach((fe) => {
+        const subItems = fe.subItems ?? []
+        visibleCount += subItems.length
+        subItems.forEach((sub) => {
+          const isSelected = [...selectedMap.values()].some((selParent) =>
+            selParent.subItems?.some(
+              (selectedSub) => selectedSub.subId === sub.subId
+            )
+          )
```

Comment:
this whole part is suboptimal due to inner loops. Can we find a data structure we can check the same thing cheaper?